Make Integer Range Optimization see through for of pattern

justinmichaud opened the pull request:
#### 2d4d8ad1273ab2997fa0736c0db51123256d2db9
<pre>
Make Integer Range Optimization see through for of pattern
<a href="https://bugs.webkit.org/show_bug.cgi?id=302192">https://bugs.webkit.org/show_bug.cgi?id=302192</a>

Reviewed by Yijia Huang.

We emit:

Branch(ArithBitOr(BooleanToNumber(Compare*), BooleanToNumber(Compare*)))

in the bytecode parser for iterator_next. This patch lets IRO see through
this and extract the cases, at least for the false side of the branch.

This removes the checks in the provided microbenchmarks.

* JSTests/microbenchmarks/for-of-vs-for-each-2.js: Added.
(console.consoleStub.testForOf):
(noInline.testForOf.testForEach):
(runTest):
(noInline.runTest):
* JSTests/microbenchmarks/for-of-vs-for-each.js: Added.
(console.consoleStub.testForOf):
(noInline.testForOf.testForEach):
(5.runTest):
(noInline.runTest):
* Source/JavaScriptCore/dfg/DFGIntegerRangeOptimizationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::isBinaryUseKind):

Canonical link: <a href="https://commits.webkit.org/303542@main">https://commits.webkit.org/303542@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/597fbcd83fa484480df8d9000f0e8e64951ca7e7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132383 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4878 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43424 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139898 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84336 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/5530 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4637 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101199 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/68463 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135329 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3889 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118578 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81989 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3775 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1187 "Found 1 new API test failure: TestWebKitAPI.ObscuredContentInsets.TopOverhangColorExtensionLayer (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83124 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/124453 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/112746 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36695 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142549 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/130891 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4548 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37284 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109577 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4630 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3922 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109753 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27899 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3441 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114852 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57824 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4602 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33213 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/163858 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4434 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68053 "Built successfully") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/42573 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/omg-tail-call-clobber-scratch-register.js.default (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4693 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4559 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/dfg/DFGIntegerRangeOptimizationPhase.cpp:
```diff
@@ -43,7 +43,7 @@ namespace {
 static constexpr bool verbose = false;
 }
-const unsigned giveUpThreshold = 50;
+const unsigned giveUpThreshold = 500000;
```
This seems like a very significant change given: 

```
                // If you hit this case for a legitimate reason, update the giveUpThreshold
                // to the smallest values that converges.
``` 
below. I assume this is left over debugging code?

justinmichaud replied:
Yes! Sorry, missed that

Constellation reviewed (commented):
Looks like EWS are red. Can you take a look?

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGIntegerRangeOptimizationPhase.cpp:
```diff
@@ -1184,84 +1184,98 @@ class IntegerRangeOptimizationPhase : public Phase {
                 bool alreadyMerged = false;
                 if (terminal->op() == Branch) {
-                    Relationship relationshipForTrue;
+                    Vector<Relationship, 1> relationshipForTrue;
+                    Vector<Relationship, 1> relationshipForFalse;
```
Could you please leave a comment `All facts in both relationshipForTrue and relationshipForFalse are ALWAYS combined with AND.` here?

justinmichaud replied:
The opposite is true in this case. They must all hold independently, and it is always correct to remove a relationship. That is why I can't also add the inverse cases (i.e, branch(BitAnd(a, b)) gives a and b on the true branch, but nothing on the false branch).

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGIntegerRangeOptimizationPhase.cpp:
```diff
@@ -1184,84 +1184,98 @@ class IntegerRangeOptimizationPhase : public Phase {
+
+                        auto [l, r] = extractBooleansFromBitwise(condition);
+
+                        if (condition->op() == ArithBitAnd) {
+                            if (auto relationship = extractRelationship(l))
+                                relationshipForTrue.append(relationship);
+                            if (auto relationship = extractRelationship(r))
+                                relationshipForTrue.append(relationship);
+                        } else if (condition->op() == ArithBitOr) {
+                            if (auto relationship = extractRelationship(l).inverse())
+                                relationshipForFalse.append(relationship);
+                            if (auto relationship = extractRelationship(r).inverse())
+                                relationshipForFalse.append(relationship);
+                        }
```
So, It's a one-level expansion of `ArithBitAnd`/`ArithBitOr` patterns.

justinmichaud replied:
I guess it expands just one half of the condition, but yes

hyjorc1 reviewed (approved):
r=me if EWS is happy.

justinmichaud committed:
Make Integer Range Optimization see through for of pattern
https://bugs.webkit.org/show_bug.cgi?id=302192

Reviewed by Yijia Huang.

We emit:

Branch(ArithBitOr(BooleanToNumber(Compare*), BooleanToNumber(Compare*)))

in the bytecode parser for iterator_next. This patch lets IRO see through
this and extract the cases, at least for the false side of the branch.

This removes the checks in the provided microbenchmarks.

* JSTests/microbenchmarks/for-of-vs-for-each-2.js: Added.
(console.consoleStub.testForOf):
(noInline.testForOf.testForEach):
(runTest):
(noInline.runTest):
* JSTests/microbenchmarks/for-of-vs-for-each.js: Added.
(console.consoleStub.testForOf):
(noInline.testForOf.testForEach):
(5.runTest):
(noInline.runTest):
* Source/JavaScriptCore/dfg/DFGIntegerRangeOptimizationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::isBinaryUseKind):

Canonical link: https://commits.webkit.org/303542@main
